src/app/examples/stock-market/stock-market.service.spec.ts

```diff
@@ -18,4 +28,44 @@ describe('StockMarketService', () => {
       expect(service).toBeTruthy();
     })
   );
+
+  it(
+    'should return expected result',
+    inject([StockMarketService], (service: StockMarketService) => {
+      const expectedStock: any = {
+        symbol: 'TSLA',
+        primaryExchange: 'Nasdaq Global Select',
+        latestPrice: 284.96,
+        change: -9.88,
+        changePercent: -0.03351
+      };
+
+      httpClientSpy.get.and.returnValue(of(expectedStock));
+
+      service
+        .retrieveStock('TS')
+        .subscribe(stock => expect(stock).toBeTruthy(), fail);
+
+      expect(httpClientSpy.get.calls.count()).toBe(1, 'called once');
+    })
+  );
+
+  it(
+    'should return error when server returns 404',
+    inject([StockMarketService], (service: StockMarketService) => {
+      const errorResponse = new HttpErrorResponse({
+        error: 'call expected error',
+        statusText: 'Not Found',
+        status: 404
+      });
+      httpClientSpy.get.and.returnValue(of(errorResponse));
+
+      service
+        .retrieveStock('TS')
+        .subscribe(
+          () => fail('expected an error'),
+          error => expect(error).not.toBeUndefined()
```

Comment:
Here I'm not sure we can test for errors, because the stock-market service tries to map the result even though the httpClient throws an error and it gives a string parsing error every time.
Looks like we should also catch the error in the stock-market service.

Reply:
I don't catch errors inside the services because it adds no real value imho. The service shouldn't invoke the `map` function because it follows the "error" flow, thus the `filter` doesn't really do anything.

For testing I think it's sufficient to handle the happy path, meaning to only test the `map` functionality. Because the error test that is written now is "just" testing some Angular internal logic, which should already be tested inside Angular and we don't have to test it ourselves.

Inside our effect we should test the `catchError` part tho.

Does this make sense 😅 ?